### PR TITLE
feat: support groupby on growing and non-indexed sealed egment(#30307)

### PR DESCRIPTION
**Pull request by MrPresent-Han**
related: #30308

### CODE REVIEW

File: internal/core/src/common/QueryResult.h [hunk 1]
```diff
@@ -45,6 +146,37 @@ struct SearchResult {
         return topk_per_nq_prefix_sum_[total_nq_];
     }
 
+ public:
```
Comment (MrPresent-Han) [inline internal/core/src/common/QueryResult.h:149]: for every query_vector, we need to assemble a VectorIterator which contains several Knowhere_iterators.
For example, if input query_vectors is 10 and 10 chunks totally.
Then, there will be 10 VectorIterators created accordingly and each of them integrates 10 Knowhere iterators.
Internally, I use priority queue to solve the sort problem inside the vectorIterator

File: internal/core/src/common/QueryResult.h [hunk 2]
```diff
@@ -31,6 +32,106 @@
 #include "knowhere/index_node.h"
 
 namespace milvus {
+
+struct OffsetDisPair {
+    std::pair<int64_t, float> off_dis_;
+    int iterator_idx_;
```
Comment (yah01) [inline internal/core/src/common/QueryResult.h:38]: the default fields visibility for struct is public, mark them `private`

File: internal/core/src/common/QueryResult.h [hunk 3]
```diff
@@ -31,6 +32,106 @@
...
+    }
+};
+struct VectorIterator {
+ public:
+    VectorIterator(int chunk_count, int64_t chunk_rows = -1)
+        : chunk_rows_(chunk_rows) {
+        iterators_.reserve(chunk_count);
+    }
+
+    std::pair<int64_t, float>
```
Comment (yah01) [inline internal/core/src/common/QueryResult.h:65]: prefer to use `std::optional<T>` then we can remove the magic value `EmptyOffsetPair`

Comment (MrPresent-Han) [reply internal/core/src/common/QueryResult.h]: changed

File: internal/core/src/common/QueryResult.h [hunk 4]
```diff
@@ -45,6 +155,37 @@ struct SearchResult {
...
+                    chunk_count, rows_per_chunk);
+                vector_iterators.emplace_back(vector_iterator);
+            }
+            auto kw_iterator = kw_iterators[i];
+            vector_iterators[vec_iter_idx++]->AddIterator(kw_iterator);
+        }
+        for (auto vector_iter : vector_iterators) {
+            vector_iter->seal();
+        }
+        this->vector_iterators_ = vector_iterators;
```
Comment (yah01) [inline internal/core/src/common/QueryResult.h:186]: this coule be moved into the field

File: internal/core/src/common/QueryResult.h [hunk 5]
```diff
@@ -31,6 +32,115 @@
...
+        }
+        return chunk_idx * chunk_rows_ + chunk_offset;
+    }
+
+ private:
+    std::vector<std::shared_ptr<knowhere::IndexNode::iterator>> iterators_;
+    std::priority_queue<std::shared_ptr<OffsetDisPair>,
+                        std::vector<std::shared_ptr<OffsetDisPair>>,
+                        OffsetDisPairComparator>
+        heap_;
```
Comment (yah01) [inline internal/core/src/common/QueryResult.h:137]: directly storing the `OffsetDisPair` may be faster as it's a small struct

File: internal/core/src/common/QueryResult.h [hunk 6]
```diff
@@ -31,6 +32,115 @@
...
+        return std::nullopt;
+    }
+    bool
+    HasNext() {
+        return !heap_.empty();
+    }
+    bool
+    AddIterator(std::shared_ptr<knowhere::IndexNode::iterator> iter) {
+        if (!sealed && iter != nullptr) {
+            iterators_.emplace_back(iter);
```
Comment (yah01) [inline internal/core/src/common/QueryResult.h:98]: move `iter` into to avoid modifying the shared pointer inner ref counter

**Pull request merged**